File: app/schemas/current.py
"""
app/schemas/current.py

Response model for GET /api/current.

Returns a list of points rather than a raw grid, since Member 1 will
render these as individual arrows/particles in Cesium (each point
needs its own position, speed, and direction — a flat list maps
directly to that, unlike the /api/model grid which is a continuous
color-mapped surface).
"""

from typing import List
from pydantic import BaseModel, Field


class CurrentPoint(BaseModel):
    latitude: float
    longitude: float
    u: float = Field(..., description="Eastward velocity component (m/s)")
    v: float = Field(..., description="Northward velocity component (m/s)")
    speed: float = Field(..., description="sqrt(u^2 + v^2), in m/s")
    direction: float = Field(..., description="Degrees clockwise from east, 0-360")


class CurrentResponse(BaseModel):
    depth: float
    time: str
    points: List[CurrentPoint]

File: src/types/current.ts
/**
 * A single ocean current sample. Unlike OceanGridData (a scalar
 * field), current data is a vector at each point — u/v components
 * plus their derived speed/direction — so it's kept as its own type
 * rather than shoehorned into OceanGridData.
 */
export interface CurrentPoint {
  latitude: number;
  longitude: number;
  /** Depth (meters) represented by this vector. */
  depth: number;
  /** ISO 8601 timestamp represented by this vector. */
  time: string;
  /** Eastward velocity component, m/s. */
  u: number;
  /** Northward velocity component, m/s. */
  v: number;
  /** Derived from u/v: sqrt(u^2 + v^2). */
  speed: number;
  /** Derived from u/v: atan2(v, u), in radians counterclockwise from east. */
  direction: number;
}

/** A set of current samples for one depth/time slice. */
export interface CurrentField {
  depth: number;
  time: string;
  points: CurrentPoint[];
}
